Issue3284 district chp

JayHuLBL opened the pull request.

ZhanweiHe committed:
Add district CHP package

This is a package for the CHP model applied in district energy system. In the package, here are: ToppingCycle, BottomingCycle, Combined, Data, BaseClasses, Validation, and Obsolete.

JayHuLBL committed:
Merge pull request #4083 from ZhanweiHe/issue3284_districtCHP

Add district CHP package

JayHuLBL committed:
cleaned the package order

JayHuLBL committed:
cleaned files

ZhanweiHe committed:
0) Reset code from upstream 1) Added .mos files for DistrictCHP.Validation.ToppingCycle, DistrictCHP.Validation.BottomingCycle_AnnualSim, and DistrictCHP.Examples.CombinedCycleCHP. 2) Updated package icons

JayHuLBL committed:
Merge pull request #4096 from ZhanweiHe/issue3284_districtCHP

Added .mos files for the districtCHP package.

JayHuLBL committed:
renamed example model [ci skip]

JayHuLBL committed:
Merge branch 'master' into issue3284_districtCHP

JayHuLBL committed:
corrected wrong html syntex

JayHuLBL committed:
corrected unit conventions

ZhanweiHe committed:
Delete package Buildings.Fluid.CHPs.DistrictCHP.Obsolete and ensure all instances have descriptive comments

JayHuLBL committed:
Merge pull request #4221 from ZhanweiHe/issue3284_districtCHP

Delete package Buildings.Fluid.CHPs.DistrictCHP.Obsolete and ensure a…

JayHuLBL committed:
Merge branch 'master' into issue3284_districtCHP

JayHuLBL committed:
improving the CHP models

JayHuLBL committed:
corrected html error

JayHuLBL committed:
corrected missing documentation

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/Validation/BottomCycle.mo:
```diff
@@ -0,0 +1,204 @@
+</ul>
+</html>", info="<html>
+<p>
+This validation model compares the proposed CHP bottoming-cycle model with the
```
@ZhanweiHe The documentation should be updated. The validation is not comparing the results with the one from TPL.

ZhanweiHe replied:
The documentation is good. We may need to update the model by including reference data from the TPL results; otherwise, it should not be included in the validation package

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/Validation/TopCycle.mo:
```diff
@@ -0,0 +1,110 @@
+exhaust temperature, for 10 gas turbines with capacities ranging from 4 MW to 35 MW.
+<p align=\"center\">
+<img alt=\"Image of efficiency curves\"
+src=\"modelica://CHP/Resources/Images/Validation/ToppingCycleEfficiencyCurves.png\"/>
```
@ZhanweiHe The image cannot be found. Do you have it anywhere?

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/package.mo:
```diff
@@ -0,0 +1,16 @@
+  extends Modelica.Icons.VariantsPackage;
+
+  annotation(Documentation(info="<html>
+This package includes models for combined heat and power (CHP) system in district energy systems. This combined-cycle CHP system consists of the topping and bottoming cycles.
```
@ZhanweiHe Can we have a more detailed instruction of how the model being developed? We do also need figures to show what the topping and bottoming cycle  are doing.

ZhanweiHe replied:
This conference paper [simbuild2024](https://publications.ibpsa.org/proceedings/simbuild/2024/papers/simbuild2024_2118.pdf) is helpful for understanding the big picture, but it does not cover steam turbine electricity.

JayHuLBL committed:
removed BOM

JayHuLBL reviewed (commented):
@ZhanweiHe See inline comments

ZhanweiHe committed:
1) Add system schematics and the governing equations to the package documentation 2)add two figures (the comparison with TPL) to the documentation of the bottoming cycle validation model

ZhanweiHe committed:
Update: 1) Add system schematics and the governing equations to the package documentation 2)add two figures (the comparison with TPL) to the documentation of the bottoming cycle validation model

ZhanweiHe committed:
Revert "1) Add system schematics and the governing equations to the package documentation 2)add two figures (the comparison with TPL) to the documentation of the bottoming cycle validation model"

This reverts commit 49403a8dab292f43402db0bcb23a2bdeb6e39307.

B
1) Add system schematics and the governing equations to the package documentation 2)add two figures (the comparison with TPL) to the documentation of the bottoming cycle validation model

ZhanweiHe committed:
Merge branch 'issue3284_districtCHP' of https://github.com/ZhanweiHe/modelica-buildings into issue3284_districtCHP

JayHuLBL committed:
Merge pull request #4297 from ZhanweiHe/issue3284_districtCHP

Update documentation

JayHuLBL committed:
deleted trash file, update imges, added link

JayHuLBL committed:
added prefix r

JayHuLBL committed:
upgraded node version

JayHuLBL committed:
upgraded node version before installation

JayHuLBL committed:
switch to the desired node version

JayHuLBL committed:
added sudo

JayHuLBL committed:
changed approach to upgrade node

JayHuLBL committed:
changed back node upgrade approach

JayHuLBL committed:
added sudo

JayHuLBL committed:
add npm installation

JayHuLBL committed:
added rehash

JayHuLBL committed:
added command end

JayHuLBL committed:
added apt update

JayHuLBL committed:
moved the node version update to the test script

JayHuLBL committed:
added npm installation

JayHuLBL committed:
updated node js version

JayHuLBL committed:
changed yml file syntax error

JayHuLBL committed:
corrected .yml error

JayHuLBL committed:
added specific node version

JayHuLBL committed:
changed to the latest node version

JayHuLBL committed:
changed to the latest node version

JayHuLBL committed:
Merge branch 'issue4306_addPrefixR' into issue3284_districtCHP

JayHuLBL committed:
merged master

JayHuLBL committed:
changed single file package to multiple files

JayHuLBL committed:
improving the documentation [ci skip]

JayHuLBL committed:
updated the figure [ci skip]

JayHuLBL committed:
Merge branch 'master' into issue3284_districtCHP

JayHuLBL committed:
Merge branch 'master' into issue3284_districtCHP

JayHuLBL committed:
clean up documentation

JayHuLBL committed:
Merge branch 'master' into issue3284_districtCHP

ViswanathanGanesh01 committed:
Updated Bottom Cycle Validation

 1. Added validations data
 2. Updated the BottomCycle validation model
 3. Updated the command script

JayHuLBL committed:
Merge pull request #4414 from ViswanathanGanesh01/Issue3284

Updated Bottom Cycle Validation

JayHuLBL committed:
removed BOM

JayHuLBL committed:
improving the format

JayHuLBL committed:
combined multiple validation data files to single file

JayHuLBL committed:
added model to exclusion list

JayHuLBL committed:
added release note

JayHuLBL committed:
removed CPU time reference data

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/BaseClasses/WasteHeatEnthalpy.mo:
```diff
@@ -0,0 +1,41 @@
+equation
+  //Waste heat specific enthalpy calculation
+  // Avaeraged specific heat of exhaust is 1.08 kJ/(kg-K)
+  hWasHea = 1.08e3*(TExh - TAmb);
```
@ViswanathanGanesh01 Could you please let me know if there is any comments on this?

JayHuLBL committed:
Merge branch 'master' into issue3284_districtCHP

JayHuLBL committed:
propagated parameters of the bottom cycle [ci skip]

JayHuLBL committed:
propagated parameters

JayHuLBL committed:
created new bottom cycle model

JayHuLBL commented:
@ZhanweiHe @ViswanathanGanesh01 

The `Buildings.Fluid.CHPs.DistrictCHP.Validation.BottomCycle` and `Buildings.Fluid.CHPs.DistrictCHP.Examples.Combined` failed to run with OpenModelica, with errors as below:
<img width="1025" height="135" alt="Screenshot 2025-12-19 at 11 21 20 AM" src="https://github.com/user-attachments/assets/3fcc924f-982f-43b9-a2d4-5994cddfffbd" />

With Dymola, running the unit test for package `Buildings.Fluid.CHPs.DistrictCHP` gives following warning:
<img width="1027" height="198" alt="Screenshot 2025-12-19 at 11 20 19 AM" src="https://github.com/user-attachments/assets/123daf31-dcb5-472d-b7d5-879d037710c7" />

The errors and warnings are related the water feeding pump in the bottom cycle. Why do we use the model `Modelica.Fluid.Machines.ControlledPump`, rather than the `Buildings.Fluid.Movers.FlowControlled_m_flow`? I created the test model `BottomCycle_test` by replacing the pump instance with the pump model from MBL. Both warnings and errors are gone. 

Please check if the changes are good to you.

JayHuLBL committed:
updated references [ci skip]

JayHuLBL committed:
merged master

JayHuLBL committed:
replaced the pump in the bottom cycle

JayHuLBL committed:
fixed html error

JayHuLBL committed:
reverted changes in conf.yml

JayHuLBL committed:
fixed typos

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/Data.mo:
```diff
@@ -0,0 +1,1087 @@
+
+  package SolarTurbines "Solar Turbines data sheets"
+
+    package NaturalGas "The fuel used is natural gas"
```
@ViswanathanGanesh01 Just wants to follow-up with this. Do we have any feedback of how to refer the data?

ViswanathanGanesh01 replied:
Hi @JayHuLBL, All the data sheet is available on the website (https://www.solarturbines.com/en_US/products/power-generation-packages.html) of Solar Turbine. so, I think we can cite that. If we need equipment level reference, I have attached the links for the datasheet.

Centaur40_4700S_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-62290
Centaur50_T6200S_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-37893
Mercury50_6400R_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150710-52396-21070
Taurus60_7901S_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-49890
Taurus65_8701S_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-38597
Taurus70_11101S_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-12880
Mars100_16000S_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-29539
Titan130_20501S_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-43744
Titan350_47000_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20220318-b6734-70be9
Titan350_53000_NG: https://s7d2.scene7.com/is/content/Caterpillar/CM20220318-18bf4-966d4
Centaur40_4700S_Diesel: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-62290
Centaur50_T6200S_Diesel: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-37893
Taurus60_7901S_Diesel: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-49890
Taurus70_11101S_Diesel: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-12880
Mars100_16000S_Diesel: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-29539
Titan130_20501S_Diesel: https://s7d2.scene7.com/is/content/Caterpillar/CM20150703-52095-43744

JayHuLBL replied:
Thanks! I just added all the links to the info section.

ZhanweiHe replied:
@ViswanathanGanesh01 thank you for adding the link. It should be noted that the detailed off-design performance data for those GTs are not available in above links

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/TopCycle.mo:
```diff
@@ -0,0 +1,221 @@
+  Buildings.Controls.OBC.CDL.Reals.MultiplyByParameter exhMas(
+    final k=per.mExh_nominal) "Exhaust mass flow rate computation"
+    annotation (Placement(transformation(extent={{-20,-90},{0,-70}})));
+  Buildings.Controls.OBC.CDL.Reals.MultiplyByParameter loaPer(
```
@ZhanweiHe Why do we need this block? Multiplying 1 basically doesn't change anything.

JayHuLBL replied:
@ViswanathanGanesh01

ZhanweiHe replied:
Dr. Hu, you are right. Please delete this `loaPer`. Previously, I used the load percentage as the input (the gain was 100), and later I changed it to a load ratio (0-1), so I changed the gain value from 100 to 1.

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/BottomCycle.mo:
```diff
@@ -0,0 +1,373 @@
+    "Water medium - Medium model for port_a (inlet)";
+
+  // Parameters for the calculation blocks
+  parameter Real a[:]={-0.23380344533,0.220477944738,-0.01476897980}
```
@ZhanweiHe How should user specify the parameter? What is the array size? Does it have to be 3, or can be arbitrary?

The same comment goes to the parameter `a_SteMas`.
@ViswanathanGanesh01

ZhanweiHe replied:
@JayHuLBL Dr. Hu, the size should be fixed as 3. The size of `a_SteMas` should be fixed as 3 as well.

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/BaseClasses/Functions/ExhaustExergyEfficiency.mo:
```diff
@@ -0,0 +1,56 @@
+
+  input Modelica.Units.NonSI.Temperature_degF TExh
+    "Exhaust gas temperature in degrees Fahrenheit";
+  input Real a[3]={0.2441, 0.0746, -0.00279}
```
Why does the input array `a` have value? And, it seems that the array size is fixed to 3? But in the top-level class that instantiates the function, the array size becomes arbitrary?
@ViswanathanGanesh01 @ZhanweiHe

ZhanweiHe replied:
@JayHuLBL Yes, the coefficient sizes for exhaust exergy efficiency should be set to 3. Dr. Hu, you are right, in the top-level class, the array size should be fixed as 3

JayHuLBL replied:
Do we need to keep the assignment to the `a[3]`? Is the `{0.2441, 0.0746, -0.00279}` a set of important value that we should keep it in certain way?

ZhanweiHe replied:
@JayHuLBL Dr. Hu, yes, please keep it as a[3]. {0.2441, 0.0746, -0.00279} is the original coefficient set from textbook https://doi.org/10.1201/9780429244360. I think it would be an example for users. They can use it as default value or input their coefficient

JayHuLBL replied:
Can this set of the coefficient be found directly from the reference, or we need to do the regression to generate it?

ZhanweiHe replied:
The approach supports both cases: if operational data is available, users can perform linear regression to obtain site-specific coefficients; otherwise, the default coefficients provide estimated steam turbine electricity output based on exhaust conditions.

JayHuLBL committed:
renaming variable names [ci skip]

JayHuLBL committed:
added data sources link [ci skip]

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/Combined.mo:
```diff
@@ -0,0 +1,221 @@
+            {-60,-60}})));
+
+  // Parameters for the calculation blocks
+  parameter Real a[:]={-0.23380344533,0.220477944738,-0.01476897980}
```
Should user change the value of parameter `a` and `a_SteMas`? If yes, do we have any reference or instruction of how to change them?

ZhanweiHe replied:
If they have their own operational data,  they can do linear regression to get their coefficients. Otherwise, they might choose to use the default values for quick implementation.

JayHuLBL replied:
Where does this set of coefficient come from? Is it directly from the ThermoPower Library, or is it based on a plot in the ThermoPower Library and do the regression?

If the user needs to do the linear regression, what data sets are needed? We may need to provide the original data and the script (assuming Python) for the regression.

ZhanweiHe replied:
The coefficients are obtained via linear regression on simulated data from the ThermoPower Library example model. The regression follows the form y = f(x₁, x₂, ...) where y is the dependent variable (e.g., electricity output) and x₁, x₂ are the independent variables (e.g., exhaust conditions). A simple Python script using numpy or sklearn can be sufficient.

ZhanweiHe replied:
@JayHuLBL Here is the python script I used 
[BotCyc_OptCoef.zip](https://github.com/user-attachments/files/28850718/BotCyc_OptCoef.zip)

JayHuLBL committed:
fixed array size

JayHuLBL commented on Buildings/Fluid/CHPs/DistrictCHP/Combined.mo:
```diff
@@ -0,0 +1,221 @@
+  parameter Real a[3]={-0.23380344533,0.220477944738,-0.01476897980}
+    "Coefficients for calculating steam turbine exhaust exergy efficiency"
+    annotation (Dialog(group="Steam turbine"));
+  parameter Real a_SteMas[3]={0.153, 0.018, 0.002}
```
@ZhanweiHe 
How to specify the `a_SteMas`? If it is the value from the linear regression, do you have example code? The code you shared previously for finding the `a` is a good one, as from that code, user knows what dataset should be pvoided to do the regression.
A similar code would be good for finding the `a_SteMas`.

ZhanweiHe replied:
@JayHuLBL The a_SteMas values {0.153, 0.018, 0.002} are actually reused directly from the textbook, so I didn't perform a regression for this one. Since the correlation format is known, we could certainly generate a similar Python script. Dr. Hu, what do you think? Is it necessary to generate a new script, or are the default textbook values sufficient for this case?

JayHuLBL replied:
Based on the textbook, should the coefficients be changed according to different operation conditions? If not, then why don't we directly use it in `DistrictCHP.BaseClasses.Functions.SteamToExhaustMassFlowRatio` and don't propagate it? If the coefficients should be changed, we need to do the same as the `a[:]`.

I also noticed in the validaiton model `Buildings.Fluid.CHPs.DistrictCHP.Validation.BottomCycle`, the `a_SteMas[:]` has different value (`{0.1140,0,0}`). How did we find the value?

ZhanweiHe replied:
The `a_SteMas[:]` coefficients are site-specific and will vary depending on operational conditions, so they do need to be propagated in the same way as `a[:]` rather than being hardcoded inside `DistrictCHP.BaseClasses.Functions.SteamToExhaustMassFlowRatio`.

Regarding the values in `Buildings.Fluid.CHPs.DistrictCHP.Validation.BottomCycle`: these were obtained from the ThermoPower library. Since the validation model only runs for 3000 seconds, the ratio of exhaust flow rate to steam flow rate remains approximately constant over that window, so I used those fixed values directly for the purpose of that example.

ZhanweiHe replied:
@JayHuLBL I replaced the correlation format. Here is the code. 
[BotCyc_OptCoef_MuRatio.py](https://github.com/user-attachments/files/28960903/BotCyc_OptCoef_MuRatio.py)
The users need exhaust gas flow, steam mass flow, exhaust temperature, and superheated steam temperature as the reference data for the regression.

JayHuLBL replied:
@ZhanweiHe Do we have any reference data for the regression to find the `a_SteMas[:]`?

ZhanweiHe replied:
@JayHuLBL No, I do not have the reference data.

JayHuLBL replied:
@ZhanweiHe Is it possible to prepare a set of data set, for the regression  so that when user runs the script `https://github.com/lbl-srg/modelica-buildings/blob/issue3284_districtCHP/Buildings/Resources/Data/Fluid/CHPs/DistrictCHP/Validation/operationCoefficients.py`, it will read into the data set and gives the coefficient `a[:]` and `a_SteMas[:]`? I tried to run the model `https://github.com/casella/ThermoPower/blob/e2b011ac7fd90f9cf5771f29f1aefa160550b6ee/ThermoPower/Examples.mo#L3119`, but it would take me much more time to retrieve the needed data set and your help would be much appreciated. It is for the `Buildings.Fluid.CHPs.DistrictCHP.Validation.BottomCycle`.

JayHuLBL replied:
@ZhanweiHe Could you please run the ThermoPower example model and prepare the needed data set for running the `operationCoefficients.py` as described above? The date set should includes `exhaust gas temperature  [degC]`, `exhaust gas mass flow rate  [kg/s]`, `power generation [W]`, `Steam temperature [degC]`, and `Steam mass flow [kg/s]`.

JayHuLBL committed:
renamed connectors [ci skip]

JayHuLBL committed:
Merge branch 'master' into issue3284_districtCHP

JayHuLBL committed:
adding script [ci skip]

JayHuLBL committed:
deleted wrongly placed files [ci skip]
